app/assets/javascripts/autolaunch.js

```diff
@@ -0,0 +1,223 @@
+function autolaunchInteractive (documentId, launchUrl) {
+  var CURRENT_VS_LINKED = "Another page contains more recent data. Which would you like to use?";
+  var LINKED_VS_LINKED = "There are two possibilities for continuing your work. Which version would you like to use?";
+
+  // Update the loading message after 10 seconds
+  var showTimeoutId = setTimeout(function() {
+    $('#loading-text').html("Still loading!  You may want to reload this page to try again.")
+  }, 10000);
+
+  var phone = iframePhone.getIFrameEndpoint();
+  // Variables below are set in `initInteractive` handler.
+  var interactiveData = null;
+  var directlyLinkedState = null;
+  var mostRecentLinkedState = null;
+  var interactiveStateAvailable = false;
+
+  function stateValid (state) {
+    return !!(state && state.docStore && state.docStore.recordid && state.docStore.accessKeys && state.docStore.accessKeys.readOnly);
+  }
+
+  function showDataSelectDialog (twoLinkedStates) {
+    function showPreview (element) {
+      $(element).addClass('preview-active');
+      $('.overlay').show();
+    }
+    function hidePreview () {
+      $('.preview-active').removeClass('preview-active');
+      $('.overlay').hide();
+    }
+    function launchInt () {
+      launchInteractive();
+      $('.data-select-dialog').remove();
+    }
+
+    // There are two supported cases. It's either the choice between most recent linked data and the current data.
+    // Or between the most recent data and data which is directly linked if given interactive doesn't have its own
+    // state yet.
+    if (!twoLinkedStates) {
+      $('#question').text(CURRENT_VS_LINKED);
+    } else {
+      $('#question').text(LINKED_VS_LINKED);
+    }
+    var state1 = mostRecentLinkedState;
+    var state2 = twoLinkedStates ? directlyLinkedState : interactiveData;
+
+    $('.data-select-dialog').show()
+    $('#state1-time').text((new Date(state1.updatedAt)).toLocaleString());
+    $('#state2-time').text((new Date(state2.updatedAt)).toLocaleString());
+    $('#state1-page-idx').text(state1.pageIndex);
+    $('#state2-page-idx').text(state2.pageIndex);
+    if (state1.pageName) {
+      $('#state1-page-name').text(' - ' + state1.pageName);
+    }
+    if (state2.pageName) {
+      $('#state2-page-name').text(' - ' + state2.pageName);
+    }
+    $('#state1-activity-name').text(state1.activityName);
+    $('#state2-activity-name').text(state2.activityName);
+
+    var prevUrl = state1.interactiveStateUrl;
+    var srcPrev = $.param.querystring(launchUrl, {launchFromLara: Base64.encode(JSON.stringify({ url: prevUrl }))});
```

Comment:
I'm not sure if it is worth making this change.  But you can use this form of launchFromLara params so CODAP would only have access to the readOnly key: 
https://github.com/concord-consortium/cloud-file-manager/blob/master/src/code/providers/lara-provider.coffee#L273-L279  This way if the preview gets messed up somehow CODAP won't accidentially save that state.  You could probably also just pull `lara_options.reporting_url` out of the state and use that.

This comes back to your question before: why don't we just pass the info directly to CODAP during a normal launch instead of asking it to make an additional request back to LARA. I think it comes back to supporting the 'reload the page' case combined with preventing them from sharing the URL with someone else.  If we put the full set of info encoded in the launchFromLara parameter, then the user could reload the document as much as they want, but they could also send the URL to someone else.  By using a URL back to LARA, the reloading will only work as long as the user is still logged into LARA, and if they send the URL to someone else it won't work.  With our recent JWT support in the portal for apps like GeniVenture we handled this a different way using timeouts and refreshing a token.